File: products/yappc/frontend/apps/api/src/database/repositories/project.repository.ts
/**
 * Project repository for database operations on projects.
 *
 * <p><b>Purpose</b><br>
 * Provides data access operations for Project entities, including specialized
 * queries for project management.
 *
 * <p><b>Usage</b><br>
 * <pre>{@code
 * const projectRepo = new ProjectRepository(prisma);
 * const projects = await projectRepo.findByWorkspace('ws-123');
 * }</pre>
 *
 * @doc.type module
 * @doc.purpose Project data access
 * @doc.layer product
 * @doc.pattern Repository
 */

import { PrismaClient, Project } from '@prisma/client';
import { BaseRepository } from '../repository.base';

/**
 * Repository for Project entity operations.
 *
 * <p><b>Purpose</b><br>
 * Provides specialized queries for project management including filtering by
 * workspace, status, and type.
 *
 * @doc.type class
 * @doc.purpose Project repository
 * @doc.layer product
 * @doc.pattern Repository
 */
export class ProjectRepository extends BaseRepository<Project, 'project'> {
  /**
   * Creates a new ProjectRepository instance.
   *
   * @param prisma - Prisma client instance
   */
  constructor(prisma: PrismaClient) {
    super(prisma, 'project');
  }

  /**
   * Finds all projects in a workspace.
   *
   * <p><b>Purpose</b><br>
   * Retrieves all projects belonging to a specific workspace.
   *
   * @param workspaceId - The workspace ID
   * @param options - Query options
   * @returns Array of projects
   *
   * @doc.type method
   * @doc.purpose Find projects by workspace
   * @doc.layer product
   * @doc.pattern Repository
   */
  async findByWorkspace(
    workspaceId: string,
    options?: { skip?: number; take?: number }
  ): Promise<Project[]> {
    return this.findMany({ workspaceId }, options);
  }

  /**
   * Finds active projects in a workspace.
   *
   * <p><b>Purpose</b><br>
   * Retrieves only active projects in a workspace.
   *
   * @param workspaceId - The workspace ID
   * @returns Array of active projects
   *
   * @doc.type method
   * @doc.purpose Find active projects
   * @doc.layer product
   * @doc.pattern Repository
   */
  async findActiveByWorkspace(workspaceId: string): Promise<Project[]> {
    return this.findMany({
      workspaceId,
      status: 'ACTIVE',
    });
  }

  /**
   * Finds projects by type.
   *
   * <p><b>Purpose</b><br>
   * Retrieves projects of a specific type.
   *
   * @param workspaceId - The workspace ID
   * @param type - The project type
   * @returns Array of projects
   *
   * @doc.type method
   * @doc.purpose Find projects by type
   * @doc.layer product
   * @doc.pattern Repository
   */
  async findByType(workspaceId: string, type: string): Promise<Project[]> {
    return this.findMany({
      workspaceId,
      type,
    });
  }

  /**
   * Counts projects in a workspace.
   *
   * <p><b>Purpose</b><br>
   * Returns the count of projects in a workspace.
   *
   * @param workspaceId - The workspace ID
   * @returns Count of projects
   *
   * @doc.type method
   * @doc.purpose Count projects
   * @doc.layer product
   * @doc.pattern Repository
   */
  async countByWorkspace(workspaceId: string): Promise<number> {
    return this.count({ workspaceId });
  }

  /**
   * Archives a project.
   *
   * <p><b>Purpose</b><br>
   * Changes a project's status to ARCHIVED.
   *
   * @param projectId - The project ID
   * @returns The archived project
   *
   * @doc.type method
   * @doc.purpose Archive project
   * @doc.layer product
   * @doc.pattern Repository
   */
  async archive(projectId: string): Promise<Project> {
    return this.update({ id: projectId }, { status: 'ARCHIVED' });
  }
}

File: products/yappc/backend/persistence/src/main/java/com/ghatana/yappc/api/repository/ProjectRepository.java
/*
 * Copyright (c) 2025 Ghatana Technologies
 * YAPPC API Module
 */
package com.ghatana.yappc.api.repository;

import com.ghatana.products.yappc.domain.model.Project;
import io.activej.promise.Promise;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Repository interface for Project persistence operations.
 *
 * @doc.type interface
 * @doc.purpose Project repository for CRUD operations
 * @doc.layer repository
 * @doc.pattern Repository
 */
public interface ProjectRepository {

    /** Save a project. */
    Promise<Project> save(Project project);

    /** Find a project by ID. */
    Promise<Optional<Project>> findById(UUID workspaceId, UUID id);

    /** Find a project by key. */
    Promise<Optional<Project>> findByKey(UUID workspaceId, String key);

    /** Find projects by workspace. */
    Promise<List<Project>> findByWorkspace(UUID workspaceId);

    /** Find active (non-archived) projects by workspace. */
    Promise<List<Project>> findActiveByWorkspace(UUID workspaceId);

    /** Search projects by name. */
    Promise<List<Project>> searchByName(UUID workspaceId, String query);

    /** Delete a project. */
    Promise<Boolean> delete(UUID workspaceId, UUID id);

    /** Check if a project exists. */
    Promise<Boolean> exists(UUID workspaceId, UUID id);

    /** Check if a project key is available. */
    Promise<Boolean> isKeyAvailable(UUID workspaceId, String key);

    /** Count projects by workspace. */
    Promise<Long> countByWorkspace(UUID workspaceId);
}
